### PR TITLE
lang (feature: #36): Add python SDK

### PR DESCRIPTION
**Pull request by Matts966**
This PR adds Python binding discussed in #36 
The directory structure follows [the Apache Beam project](https://github.com/apache/beam/tree/master/sdks/python).

## Example

```sh
❯❯❯❯❯ pip3.11 install -e "git+https://github.com/Matts966/wvlet/#egg=wvlet&subdirectory=sdks/python"
```

```python
>>> from wvlet import runner
>>> wv = runner.WvletRunner()
>>> sql = wv.compile("from lineitem limit 10")
-- wvlet version=2024.9.12, src=01JD172WDJ83V73FQ4QV85CQ79.wv:1
select * from lineitem
limit 10
>>> print(sql)
select * from lineitem
limit 10

```

## Tasks

- [x] Add GitHub installable python package
- [x] Install `wvlet` correctly and test it
- [x] Add [dynamic versioning](https://packaging.python.org/en/latest/discussions/single-source-version/#single-source-version), used calendar versioning
- [x] Add `target` parameter

## Tasks not addressed

- [ ] Install `wvc` automatically on `pip install` (might be addressed in future PRs)
- [ ] Upload the lib to PyPI (might be addressed in future PRs because it requires an account for the Wvlet project)
- [ ] Placeholder support (not needed immediately?)

### CODE REVIEW

**Comment by xerial** [comment]
- We can use the "wvlet" module name on PyPI https://pypi.org/project/wvlet/ (which I created a long time ago).
- For the first milestone, let's focus on compiling a single Wvlet query to SQL texts using the `wvlet compile -t (target DBMS)` command. 
    - example: PRQL python module https://github.com/PRQL/pyprql?tab=readme-ov-file#compilation
    - Once SQL can be generated, we can leverage the existing ecosystem for running SQL queries  
    - Native binaries of the compiler (wvc) aren't ready yet, as we still need to figure out how to set up cross-compilation #471.
- Let's assume the `wvlet` command is installed locally. We can also assume Java is installed on the local machine (as in PySpark).
    - Until the wvc binaries are ready, we can bundle wvlet-cli.tar.gz (or download it from the GitHub release page) and extract it when running the Python module.
- Use the calendar versioning scheme 2024.9.x for consistency. Python is also considering adopting calendar versioning https://peps.python.org/pep-2026/#calendar-versioning. #170 discuss the details on why Wvlet uses YYYY.(milestone month).(patch) versioning scheme

**Comment by Matts966** [comment]
@xerial Thank you for your comment! Updated the change 🙏

File: sdks/python/wvlet/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__version__ = "v2024.11.19"
```
Comment (xerial) [inline sdks/python/wvlet/__init__.py]: Wvlet's versioning scheme is not based on the current calendar day: YYYY.(milestone month).(patch). 2024.9.13 is the next version https://github.com/wvlet/wvlet/releases

File: sdks/python/wvlet/runner.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import sys
+from typing import Optional
+import shutil
+import subprocess
+
+
+class WvletRunner():
```
Comment (xerial) [inline sdks/python/wvlet/runner.py]: This class should be WvletCompiler (compiler.py)

**Comment by xerial** [comment]
Thanks. To complete the initial version, address the following: 
 
- [ ] Rename WvletRunner -> WvletCompiler
- [ ] Add package/method documentation to set the coding standard in an earlier phase. Here is an example in Beam:  https://github.com/apache/beam/blob/master/sdks/python/apache_beam/__init__.py
- [ ] Add Apache Software Licence 2.0 header. If IntelliJ IDEA is used, the license header setting can be added in the .idea folder 
- [ ] Add a sanity test (e.g., just creating WvletCompiler object without running wvlet compile) 
- [ ] Add documentation (e.g., website/docs/sdks/python.md) file for installation and usage instruction. You can tell that Python SDK is still beta by using https://docusaurus.io/docs/markdown-features/admonitions, e.g., 
```
:::info
Python SDK is currently in beta, requiring wvlet and Java installation on your local machine.
:::
```
- [ ] Add a GitHub Action workflow (sdks.yml) to test the behavior. 
   - [ ] (stretch) Build wvlet compiler, then build python SDK and run tests
- [ ] Set the current module version from Git tags. Here is an example script to read the current version: https://github.com/wvlet/airframe/blob/main/scripts/dynver.sh. Wvlet uses https://github.com/sbt/sbt-dynver for automatic versioning 


We can split these tasks into multiple PRs. It's up to you.

**Comment by Matts966** [comment]
@xerial Thank you for your review!
For dynamic versioning based on git tags, how about setting the version when building wheels in the future PR?
Like, `echo $GIT_TAG > __init__.py` on the GitHub Actions script.
It's straightforward and compared to external tools like [setuptools-scm](https://setuptools-scm.readthedocs.io/en/latest/), it won't break `pip install git+...`.
I removed the hard-coded version from `__init__.py` for that.

**Comment by Matts966** [comment]
@xerial Thank you for today's discussion. I modified the code. Please take a look 🙏 

- [x] Rename WvletRunner -> WvletCompiler
- [x] Add package/method documentation to set the coding standard in an earlier phase.
- [x] Add Apache Software Licence 2.0 header. If IntelliJ IDEA is used, the license header setting can be added in the .idea folder
- [x] Add a sanity test (e.g., just creating WvletCompiler object without running wvlet compile)
- [x] Add documentation (e.g., website/docs/development/python.md) file for installation and usage instruction.
- [x] Add a GitHub Action workflow (sdks.yml) to test the behavior.

File: .github/workflows/sdks.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+on: [push]
+
+jobs:
+  python:
+
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["pypy3.10", "3.9", "3.10", "3.11", "3.12", "3.13"]
```
Comment (xerial) [inline .github/workflows/sdks.yaml]: As GitHub Action is not limitless and supporting multiple Python versions will be cumbersome, let's support only the latest version. 
```suggestion
        python-version: ["3.13"]
```

File: .github/workflows/sdks.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: unittest
+        run: cd sdks/python && python -m unittest
```
Comment (xerial) [inline .github/workflows/sdks.yaml]:
```suggestion
        run: python -m unittest
        working-directory: sdks/python
```

File: .github/workflows/sdks.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+name: SDKs
+
+on: [push]
```
Comment (xerial) [inline .github/workflows/sdks.yaml]: This won't trigger build for pull requests. Use the following:
```suggestion
on:
  pull_request:
  push:
    branches:
      - main
```

**Comment by xerial** [comment]
I need to make several changes to this PR, but as the PR is created on the main branch, so I couldn't push my changes to this PR. See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork for allowing the maintainer to modify the PR. The main branch is usually protected from direct commit. 

So I created another PR to include my fixes #524  I'll close this PR

**Pull request closed without merging**